include/sys/avl.h

```diff
@@ -259,6 +259,17 @@ extern void avl_add(avl_tree_t *tree, void *node);
  */
 extern void avl_remove(avl_tree_t *tree, void *node);
 
+/*
+ * Reinsert a node only if its order has changed relative to its nearest
+ * neighbors. To optimize performance avl_update_lt() checks only the previous
+ * node and avl_update_gt() checks only the next node. Use avl_update_lt() and
+ * avl_update_gt() only if you know the direction in which the order of the
+ * node may change.
+ */
+extern boolean_t avl_update(avl_tree_t *, void *);
```

Comment:
Seems useful!  Similar to how FreeBSD was done, it would be great to start with some smaller PR's that reorganize / add to the existing OpenZFS code to make it more suitable for OSX integration.